src/replication/replication_source_db_copy.cpp

```diff
@@ -0,0 +1,643 @@
+      {
+	heap_extract_worker_task *task = new heap_extract_worker_task (*this, class_oid);
+	inc_extract_running_thread ();
+	cubthread::get_manager ()->push_task (m_heap_extract_workers_pool, task);
+      }
+
+    /* wait for all heap extract threads to end */
+    while (get_extract_running_thread () > 0)
+      {
+	if (m_is_stop)
+	  {
+	    return;
+	  }
+	thread_sleep (10);
+      }
+    execute_and_transit_phase (HEAP_COPY_FINISHED);
+    pack_and_add_end_of_extract_heap ();
+
+    execute_and_transit_phase (SCHEMA_APPLY_TRIGGERS);
+    execute_and_transit_phase (SCHEMA_APPLY_INDEXES);
+    /* wait for indexes and triggers schema */
+    if (wait_send_triggers_indexes () == ER_INTERRUPTED)
+      {
+	return;
+      }
+
+    pack_and_add_end_of_copy ();
+
+    if (wait_slave_receive_ack (chn) != NO_ERROR)
+      {
+	return;
+      }
+  }
+
+  int source_copy_context::setup_copy_protocol (cubcomm::channel &chn)
+  {
+    UINT64 pos = 0, expected_magic;
+    std::size_t max_len = sizeof (UINT64);
+    cubstream::stream_position online_repl_pos;
+
+    if (chn.recv ((char *) &expected_magic, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+
+    if (expected_magic != replication_node::SETUP_COPY_REPLICATION_MAGIC)
+      {
+	er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::setup_copy_protocol error in setup protocol");
+	assert (false);
+	return ER_FAILED;
+      }
+
+    if (chn.send ((char *) &replication_node::SETUP_COPY_REPLICATION_MAGIC, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+
+    /* TODO */
+    online_repl_pos = m_online_replication_start_pos;
+    pos = htoni64 (online_repl_pos);
+    if (chn.send ((char *) &pos, max_len) !=  css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::setup_copy_protocol online_repl_pos:%lld",
+			      online_repl_pos);
+
+    return NO_ERROR;
+  }
+
+  int source_copy_context::wait_slave_receive_ack (cubcomm::channel &chn)
+  {
+    UINT64 pos = 0, expected_magic;
+    std::size_t max_len = sizeof (UINT64);
+
+    if (chn.recv ((char *) &expected_magic, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
+
+    if (expected_magic != replication_node::SETUP_COPY_END_REPLICATION_MAGIC)
+      {
+	er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::setup_copy_protocol error in setup protocol");
+	assert (false);
+	return ER_FAILED;
+      }
+    er_log_debug_replication (ARG_FILE_LINE, "source_copy_context::wait_slave_receive_ack OK");
+
+    return NO_ERROR;
+  }
+
+  void source_copy_context::stop ()
+  {
+    m_is_stop = true;
+    m_state_cv.notify_all();
+  }
+
+  /*
+   * utility functions for source server extraction of database replication copy
+   *
+   */
+
+  int convert_to_last_representation (cubthread::entry *thread_p, RECDES &rec_des, record_descriptor &record,
```

Comment:
Do you think that "heap extract for copydb" functionality could be kept separate of copydb source?

On first sight, it looks like a reusable functionality (e.g. unload).

Reply:
copy_class function uses replication stream, it would require some plugins objects to make it generic; I prefer to leave this for future if it proves common with other functionality.
convert_to_last_representation seems to small and specific; I prefer the same resolution.